include/nbl/builtin/hlsl/workgroup/shared_scan.hlsl

```diff
@@ -0,0 +1,184 @@
+        while(lastInvocationInLevel >= glsl::gl_SubgroupSize() * glsl::gl_SubgroupSize())
+        {
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            if(participate)
+            {
+                if (any(bool2(gl_LocalInvocationIndex == lastInvocationInLevel, isLastSubgroupInvocation)))
+                { // only invocations that have the final value of the subgroupOp (inclusive scan) store their results
+                    sharedAccessor.main.set(nextLevelStoreIndex, scan);
+                }
+            }
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
+            if(participate)
+            {
+                const uint prevLevelScan = sharedAccessor.main.get(gl_LocalInvocationIndex);
+                scan = subgroupOp(prevLevelScan);
+                if(isScan)
+                {
+                    sharedAccessor.main.set(scanStoreIndex, scan);
+                }
+            }
+            if(isScan)
+            {
+                scanStoreIndex += lastInvocationInLevel + 1u;
+            }
+        }
+        if(lastInvocationInLevel >= glsl::gl_SubgroupSize())
+        {
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            if(participate)
+            {
+                if(any(bool2(gl_LocalInvocationIndex == lastInvocationInLevel, isLastSubgroupInvocation))) {
+                    sharedAccessor.main.set(nextLevelStoreIndex, scan);
+                }
+            }
+            sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
+            participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
+            if(participate)
+            {
+                const uint prevLevelScan = sharedAccessor.main.get(gl_LocalInvocationIndex);
+                scan = subgroupOp(prevLevelScan);
+                if(isScan) {
+                    sharedAccessor.main.set(scanStoreIndex, scan);
+                }
+            }
+        }
+        return scan;
+    }
+};
+
+template<typename T, class Binop, class SharedAccessor>
+struct WorkgroupScanTail
+{
+    bool isExclusive;
+    T identity;
+    T firstLevelScan;
+    uint lastInvocation;
+    uint scanStoreIndex;
+
+    static WorkgroupScanTail create(bool isExclusive, T identity, T firstLevelScan, uint lastInvocation, uint scanStoreIndex)
+    {
+        WorkgroupScanTail<T, Binop, SharedAccessor> wst;
+        wst.isExclusive = isExclusive;
+        wst.identity = identity;
+        wst.firstLevelScan = firstLevelScan;
+        wst.lastInvocation = lastInvocation;
+        wst.scanStoreIndex = scanStoreIndex;
+        return wst;
+    }
+
+    T operator()()
+    {
+        Binop binop;
+        SharedAccessor sharedAccessor;
+        sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
```

Comment:
tail comes after head, and in the head produces a series of `gl_LocalInvocationIndex<gl_SubgroupCount` values which do not actually need to be stored to SMEM, just add an `in T headResult` and comment that ceratin lane values can be garbage (when `gl_LocalInvocation>=SubgroupCount`)

Then you also don't need this first memory barrier